lib/ecto/query/planner.ex

```diff
@@ -1420,7 +1420,7 @@ defmodule Ecto.Query.Planner do
     exprs =
       case operation do
         :all -> @all_exprs
-        :update_all -> @update_all_exprs
```

Comment:
There is no reason to change this line AFAIK? Update all already includes with_cte.

Reply:
Or do you mean that without this change ctes are not processed at all? If so, I think what we should do is, when we process ctes, the operation should always be :all. The same for subwueries.

Reply:
To be clear, pass :all instead of operation here: https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/query/planner.ex#L761

Please do the same for when we plan subqueries and combinators. :)

Reply:
Actually, I figured out other improvements in this area, so I push a commit to master. You should rebase and remove those changes and everything should still work. :heart:

Reply:
Thanks @josevalim, I've rebased and added a failing test to demonstrate the problem I'm getting.

Reply:
After taking another look, I think I found the problem. `merge_cache` and `validate_and_increment` both had to be updated to use `:all` when calling `traverse_exprs` on a CTE query expression.